modules/images/dominant-color/load.php

```diff
@@ -0,0 +1,340 @@
+<?php
+/**
+ * Module Name: Dominant Color
+ * Description: Adds support to store dominant-color for an image and create a placeholder background with that color.
+ * Experimental: Yes
+ *
+ * @package performance-lab
+ * @since n.e.x.t
+ */
+
+/**
+ * Add the dominant color metadata to the attachment.
+ *
+ * @since n.e.x.t
+ *
+ * @param array $metadata        The attachment metadata.
+ * @param int   $attachment_id   The attachment ID.
+ * @return array $metadata The attachment metadata.
+ */
+function dominant_color_metadata( $metadata, $attachment_id ) {
+	if ( ! wp_attachment_is_image( $attachment_id ) ) {
+		return $metadata;
+	}
+
+	$dominant_color = dominant_color_get_dominant_color( $attachment_id );
+	if ( ! is_wp_error( $dominant_color ) && ! empty( $dominant_color ) ) {
+		$metadata['dominant_color'] = $dominant_color;
+	}
+
+	$has_transparency = dominant_color_has_transparency( $attachment_id );
+	if ( ! is_wp_error( $has_transparency ) ) {
+		$metadata['has_transparency'] = $has_transparency;
+	}
+
+	return $metadata;
+}
+add_filter( 'wp_generate_attachment_metadata', 'dominant_color_metadata', 10, 2 );
+
+/**
+ * Filter various image attributes to add the dominant color to the image
+ *
+ * @since n.e.x.t
+ *
+ * @param array  $attr Attributes for the image markup.
+ * @param object $attachment Image attachment post.
+ * @return mixed $attr Attributes for the image markup.
+ */
+function dominant_color_update_attachment_image_attributes( $attr, $attachment ) {
+	$image_meta = wp_get_attachment_metadata( $attachment->ID );
+	if ( ! is_array( $image_meta ) ) {
+		return $attr;
+	}
+
+	if ( isset( $image_meta['has_transparency'] ) ) {
+		$attr['data-has-transparency'] = $image_meta['has_transparency'] ? 'true' : 'false';
+
+		$class = $image_meta['has_transparency'] ? 'has-transparency' : 'not-transparent';
+		if ( empty( $attr['class'] ) ) {
+			$attr['class'] = '';
+		}
+		$attr['class'] .= $class;
+	}
+
+	if ( ! empty( $image_meta['dominant_color'] ) ) {
+		$attr['data-dominant-color'] = $image_meta['dominant_color'];
+		if ( empty( $attr['style'] ) ) {
+			$attr['style'] = '';
+		}
+		$attr['style'] .= '--dominant-color: #' . $image_meta['dominant_color'] . ';';
+	}
+
+	return $attr;
+}
+add_filter( 'wp_get_attachment_image_attributes', 'dominant_color_update_attachment_image_attributes', 10, 2 );
+
+/**
+ * Filter image tags in content to add the dominant color to the image.
+ *
+ * @since n.e.x.t
+ *
+ * @param string $filtered_image The filtered image.
+ * @param string $context The context of the image.
+ * @param int    $attachment_id The attachment ID.
+ *
+ * @return string image tag
+ */
+function dominant_color_img_tag_add_dominant_color( $filtered_image, $context, $attachment_id ) {
+
+	if ( 'the_content' !== $context ) {
```

Comment:
Let's add a comment why this is here.
```suggestion
	// Only apply this in `the_content` for now, since otherwise it can result in duplicate runs due to a problem with full site editing logic.
	if ( 'the_content' !== $context ) {
```